source/uploader.ts

```diff
@@ -0,0 +1,529 @@
+    } else {
+      const { url, headers } = this.uploadMetadata;
+      this.singlePartUpload({ url, headers });
+    }
+  }
+
+  /** Create component entity and get upload metadata. */
+  async uploadPreflight() {
+    logger.debug("Registering component and fetching upload metadata.");
+
+    const component = {
+      ...this.data,
+      id: this.componentId,
+      name: this.fileName,
+      file_type: this.fileType,
+      size: this.fileSize,
+    };
+    const response = await this.session.call<
+      [CreateResponse, GetUploadMetadataResponse]
+    >([
+      operation.create("FileComponent", component),
+      {
+        action: "get_upload_metadata",
+        file_name: `${this.fileName}${this.fileType}`,
+        file_size: this.fileSize,
+        component_id: this.componentId,
+        parts: this.numParts,
+      },
+    ]);
+
+    this.createComponentResponse = response[0];
+    this.uploadMetadata = response[1];
+  }
+
+  /** Upload file using single-part upload. */
+  async singlePartUpload({
+    url,
+    headers,
+  }: {
+    url: string;
+    headers: Record<string, string>;
+  }) {
+    try {
+      await this.uploadFile({ url, headers });
+      await this.completeUpload();
+      logger.debug("Upload complete", this.componentId);
+    } catch (error) {
+      if (this.onError) {
+        this.onError(error as Error);
+      }
+    }
+  }
+
+  /** Handle progress event for single-part upload */
+  handleSinglePartProgress(
+    progressEvent: ProgressEvent<XMLHttpRequestEventTarget>
+  ) {
+    let progress = 0;
+
+    if (progressEvent.lengthComputable) {
+      progress = Math.floor((progressEvent.loaded / progressEvent.total) * 100);
+    }
+    if (this.onProgress) {
+      this.onProgress(progress);
+    }
+  }
+
+  /** Recursively upload next chunk for multi-part upload. Calls complete upload when done. */
+  uploadNextChunk(retry = 0) {
+    const activeConnections = Object.keys(this.activeConnections).length;
+
+    if (activeConnections >= this.maxConcurrentConnections) {
+      return;
+    }
+
+    if (!this.parts.length) {
+      if (!activeConnections) {
+        this.completeUpload();
+      }
+
+      return;
+    }
+
+    const part = this.parts.pop();
+    if (this.file && part) {
+      const sentSize = (part.part_number - 1) * this.chunkSize;
+      const chunk = this.file.slice(sentSize, sentSize + this.chunkSize);
+
+      const onUploadChunkStart = () => {
+        this.uploadNextChunk();
+      };
+
+      this.uploadChunk(chunk, part, onUploadChunkStart)
+        .then(() => {
+          this.uploadNextChunk();
+        })
+        .catch((error) => {
+          if (retry <= 6) {
+            retry++;
+
+            // Exponential back-off retry before giving up
+            logger.warn(
+              `Part#${part.part_number} failed to upload, backing off ${
+                2 ** retry * 100
+              } before retrying...`
+            );
+            wait(2 ** retry * 100).then(() => {
```

Comment:
Intuitively, this feels short. Did you base this on anything?

Reply:
I've replaced this with a separate implementation in back_off.

It will wait a delay of 0.5s, then double that for each attempt up to a maximum of 64 seconds. It will attempt a maximum of 16 tries before giving up.